File: tests/test_functions.cpp
#include <math.h>

#include <cmath>
#include <memory>

#include <catch2/catch_test_macros.hpp>
#include <numalgs/functions.hpp>

TEST_CASE("test_polynomial") {
  const auto& coefficients = std::vector{1.0, 0.0, 0.5};
  functions::Polynomial p(coefficients);

  REQUIRE(p(1.0) == 1.0 + 0.5);
  REQUIRE(p(2.0) == 1.0 + 0.5 * 2.0 * 2.0);
  REQUIRE(p(3.0) == 1.0 + 0.5 * 3.0 * 3.0);
}

TEST_CASE("test_composition") {
  // expected expression is y = 2*(2*x - 0.5) + 0.5 = 4*x -0.5
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{0.5, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{-0.5, 2.0});

  const auto& c = functions::ComposedFunction(p, q);
  const double expected = 4 * 0.1 - 0.5;
  const double actual = c(0.1);
  REQUIRE(expected == actual);
}

TEST_CASE("test_exponential") {
  const functions::Exponential exponential;
  const double expected = std::exp(5.0);
  const double actual = exponential(5.0);
  REQUIRE(fabs(expected - actual) < 1e-12);
}

TEST_CASE("test_exponential_composition") {
  const auto& exponential = std::make_shared<const functions::Exponential>();
  const auto& polynomial =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 2.0});

  const auto& composed = functions::ComposedFunction(exponential, polynomial);
  const double expected = std::exp((*polynomial)(3.0));
  const double actual = composed(3.0);
  REQUIRE(fabs(expected - actual) < 1e-12);
}

TEST_CASE("test_trigonometric_functions") {
  const functions::Sin sin;
  const functions::Cos cos;
  const functions::Tan tan;

  REQUIRE(fabs(std::sin(0.5) - sin(0.5)) < 1e-12);
  REQUIRE(fabs(std::cos(0.5) - cos(0.5)) < 1e-12);
  REQUIRE(fabs(std::tan(0.5) - tan(0.5)) < 1e-12);
}

TEST_CASE("test_trigonometric_composition") {
  const auto& sin = std::make_shared<const functions::Sin>();
  const auto& cos = std::make_shared<const functions::Cos>();
  const auto& tan = std::make_shared<const functions::Tan>();
  const auto& polynomial =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 2.0});

  const auto& sin_composed = functions::ComposedFunction(sin, polynomial);
  const auto& cos_composed = functions::ComposedFunction(cos, polynomial);
  const auto& tan_composed = functions::ComposedFunction(tan, polynomial);

  REQUIRE(fabs(std::sin((*polynomial)(0.5)) - sin_composed(0.5)) < 1e-12);
  REQUIRE(fabs(std::cos((*polynomial)(0.5)) - cos_composed(0.5)) < 1e-12);
  REQUIRE(fabs(std::tan((*polynomial)(0.5)) - tan_composed(0.5)) < 1e-12);
}

TEST_CASE("test_addition") {
  // expected expression is y = 2*x - 0.5 + 2*x + 0.5 = 4*x
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{0.5, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{-0.5, 2.0});

  const auto& c = functions::AddFunctions(p, q);
  const double expected = 4 * 0.1;
  const double actual = c(0.1);
  REQUIRE(fabs(expected - actual) < 1e-12);
}

TEST_CASE("test_multiplication") {
  // expected expression is y = (2*x - 0.5) * (2*x + 0.5) = 4*x^2 - 0.25
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{0.5, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{-0.5, 2.0});

  const auto& c = functions::multiply_functions(p, q);
  const double expected = 4 * 0.1 * 0.1 - 0.25;
  const double actual = (*c)(0.1);
  REQUIRE(fabs(expected - actual) < 1e-12);
  REQUIRE(std::dynamic_pointer_cast<const functions::Polynomial>(c) != nullptr);
}

TEST_CASE("test_generic_multiplication") {
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{0.5, 2.0});
  const auto& e = std::make_shared<const functions::Exponential>();

  const auto& c = functions::multiply_functions(p, e);
  const double expected = (*p)(0.1) * (*e)(0.1);
  const double actual = (*c)(0.1);
  REQUIRE(fabs(expected - actual) < 1e-12);
  REQUIRE(std::dynamic_pointer_cast<const functions::MultiplyFunctions>(c) !=
          nullptr);
}

TEST_CASE("test_subtraction") {
  // expected expression is y = 2*x + 0.5 - (2*x - 0.5) = 1
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{0.5, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{-0.5, 2.0});

  const auto& c = functions::subtract_functions(p, q);
  const double expected = 1.0;
  const double actual = (*c)(0.1);
  REQUIRE(fabs(expected - actual) < 1e-12);
}

TEST_CASE("test_division") {
  // expected expression is y = (2*x + 1) / (x + 1)
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 1.0});

  const auto& c = functions::divide_functions(p, q);
  const double expected = (2.0 * 0.5 + 1.0) / (0.5 + 1.0);
  const double actual = (*c)(0.5);
  REQUIRE(fabs(expected - actual) < 1e-12);
}

TEST_CASE("test_polynomial_derivative") {
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 2.0, 3.0});
  const auto& dp = functions::derivative(p);

  REQUIRE(fabs((*dp)(2.0) - 14.0) < 1e-12);
  REQUIRE(std::dynamic_pointer_cast<const functions::Polynomial>(dp) != nullptr);

  const auto& constant =
      std::make_shared<const functions::Polynomial>(std::vector{5.0});
  const auto& d_constant = functions::derivative(constant);
  REQUIRE(fabs((*d_constant)(2.0)) < 1e-12);
}

TEST_CASE("test_basic_function_derivatives") {
  const auto& exponential = std::make_shared<const functions::Exponential>();
  const auto& sin = std::make_shared<const functions::Sin>();
  const auto& cos = std::make_shared<const functions::Cos>();
  const auto& tan = std::make_shared<const functions::Tan>();

  REQUIRE(fabs((*functions::derivative(exponential))(0.5) - std::exp(0.5)) <
          1e-12);
  REQUIRE(fabs((*functions::derivative(sin))(0.5) - std::cos(0.5)) < 1e-12);
  REQUIRE(fabs((*functions::derivative(cos))(0.5) + std::sin(0.5)) < 1e-12);
  REQUIRE(fabs((*functions::derivative(tan))(0.5) -
               1.0 / (std::cos(0.5) * std::cos(0.5))) < 1e-12);
}

TEST_CASE("test_expression_derivatives") {
  const double x = 0.5;
  const auto& p =
      std::make_shared<const functions::Polynomial>(std::vector{1.0, 2.0});
  const auto& q =
      std::make_shared<const functions::Polynomial>(std::vector{3.0, 4.0});
  const auto& exponential = std::make_shared<const functions::Exponential>();
  const auto& sin = std::make_shared<const functions::Sin>();

  const auto& add = functions::add_functions(p, exponential);
  REQUIRE(fabs((*functions::derivative(add))(x) - (2.0 + std::exp(x))) <
          1e-12);

  const auto& subtract = functions::subtract_functions(p, exponential);
  REQUIRE(fabs((*functions::derivative(subtract))(x) - (2.0 - std::exp(x))) <
          1e-12);

  const auto& multiply = functions::multiply_functions(p, exponential);
  REQUIRE(fabs((*functions::derivative(multiply))(x) -
               (2.0 * std::exp(x) + (*p)(x) * std::exp(x))) < 1e-12);

  const auto& divide = functions::divide_functions(p, q);
  const double expected_divide = (2.0 * (*q)(x) - (*p)(x) * 4.0) /
                                 ((*q)(x) * (*q)(x));
  REQUIRE(fabs((*functions::derivative(divide))(x) - expected_divide) < 1e-12);

  const auto& composed =
      std::make_shared<const functions::ComposedFunction>(sin, p);
  REQUIRE(fabs((*functions::derivative(composed))(x) -
               std::cos((*p)(x)) * 2.0) < 1e-12);
}

File: pytests/test_functions.py
import sys
sys.path.append("build")

import math
from helpers import assert_relative_close
import numalgs_py

def test_polynomial():
    slope = 0.5
    intercept = 1.0
    p = numalgs_py.Polynomial([intercept,slope])
    assert_relative_close(p(0.1), slope * 0.1 + 1.0)

def test_polynomial_str():
    p = numalgs_py.Polynomial([1.0, 0.0, 2.0])
    p_str = str(p)
    assert p_str == "1.000000 +2.000000*x^2"

def test_compose_functions():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert_relative_close((p1(p2))(2.0), p1(p2(2.0)))

def test_exponential():
    e = numalgs_py.Exponential()

    assert_relative_close(e(5.0), math.exp(5.0))

def test_exponential_str():
    e = numalgs_py.Exponential()

    assert str(e) == "exp(x)"

def test_exponential_composition():
    e = numalgs_py.Exponential()
    p = numalgs_py.Polynomial([1.0, 2.0])

    assert_relative_close(e(p)(3.0), math.exp(p(3.0)))

def test_trigonometric_functions():
    sin = numalgs_py.Sin()
    cos = numalgs_py.Cos()
    tan = numalgs_py.Tan()

    assert_relative_close(sin(0.5), math.sin(0.5))
    assert_relative_close(cos(0.5), math.cos(0.5))
    assert_relative_close(tan(0.5), math.tan(0.5))

def test_trigonometric_functions_str():
    sin = numalgs_py.Sin()
    cos = numalgs_py.Cos()
    tan = numalgs_py.Tan()

    assert str(sin) == "sin(x)"
    assert str(cos) == "cos(x)"
    assert str(tan) == "tan(x)"

def test_trigonometric_composition():
    sin = numalgs_py.Sin()
    cos = numalgs_py.Cos()
    tan = numalgs_py.Tan()
    p = numalgs_py.Polynomial([1.0, 2.0])

    assert_relative_close(sin(p)(0.5), math.sin(p(0.5)))
    assert_relative_close(cos(p)(0.5), math.cos(p(0.5)))
    assert_relative_close(tan(p)(0.5), math.tan(p(0.5)))

def test_add_functions():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert_relative_close((p1 + p2)(2.0), p1(2.0) + p2(2.0))

def test_add_polynomials_types():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert type(p1 + p2) == type(p1)
    assert type(p1(p2)) != type (p1)    # we may want to change this

def test_mult_functions():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert_relative_close((p1 * p2)(2.0), p1(2.0) * p2(2.0))
    assert type(p1 * p2) == type(p1)

def test_subtract_functions():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert_relative_close((p1 - p2)(2.0), p1(2.0) - p2(2.0))
    assert type(p1 - p2) == type(p1)

def test_divide_functions():
    p1 = numalgs_py.Polynomial([1.0, 0.5])
    p2 = numalgs_py.Polynomial([0.7, 0.3])

    assert_relative_close((p1 / p2)(2.0), p1(2.0) / p2(2.0))
    assert type(p1 / p2) != type(p1)

def test_polynomial_derivative():
    p = numalgs_py.Polynomial([1.0, 2.0, 3.0])

    assert_relative_close(p.derivative()(2.0), 14.0)
    assert_relative_close(numalgs_py.derivative(p)(2.0), 14.0)

    constant = numalgs_py.Polynomial([5.0])
    assert_relative_close(constant.derivative()(2.0), 0.0)

def test_basic_function_derivatives():
    exponential = numalgs_py.Exponential()
    sin = numalgs_py.Sin()
    cos = numalgs_py.Cos()
    tan = numalgs_py.Tan()

    assert_relative_close(exponential.derivative()(0.5), math.exp(0.5))
    assert_relative_close(sin.derivative()(0.5), math.cos(0.5))
    assert_relative_close(cos.derivative()(0.5), -math.sin(0.5))
    assert_relative_close(tan.derivative()(0.5), 1.0 / (math.cos(0.5) ** 2))

def test_expression_derivatives():
    x = 0.5
    p = numalgs_py.Polynomial([1.0, 2.0])
    q = numalgs_py.Polynomial([3.0, 4.0])
    exponential = numalgs_py.Exponential()
    sin = numalgs_py.Sin()

    assert_relative_close((p + exponential).derivative()(x), 2.0 + math.exp(x))
    assert_relative_close((p - exponential).derivative()(x), 2.0 - math.exp(x))
    assert_relative_close(
        (p * exponential).derivative()(x),
        2.0 * math.exp(x) + p(x) * math.exp(x),
    )

    expected_divide = (2.0 * q(x) - p(x) * 4.0) / (q(x) * q(x))
    assert_relative_close((p / q).derivative()(x), expected_divide)
    assert_relative_close(sin(p).derivative()(x), math.cos(p(x)) * 2.0)
